docs/user_guide/ml.md

```diff
@@ -1,8 +1,264 @@
+Add in your `law.cfg`, under the sections `analysis`, a `ml_modules` keyword, where you point to the Python module where the model's definition and derivation happens.
+These import structures are relative to the analysis root directory.
+```bash
+[analysis]
+# any other config entries
+
+# if you want to include multiple things from the same parent module, you can use a comma-separated list in {}
+ml_modules: hbt.ml.{test}
+inference_modules: hbt.inference.test
+```
+# ABC functions
+In the following we will go through several abstract functions that you must overwrite, in order to be able to use your custom ML class with columnflow.
+
+## sandbox:
+In {py:meth}`~columnflow.ml.MLModel.sandbox`, you specify which sandbox setup file should be sourced to setup the environment for ML usage.
+The return value of {py:meth}`~columnflow.ml.MLModel.sandbox` is the path to your shell (sh) file, e.g:
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def sandbox
+:end-at: return "bash::$HBT_BASE/sandboxes/venv_columnar_tf.sh"
+```
+It is recommended to start the path with `bash::`, to indicate that you want to source the {py:meth}`~columnflow.ml.MLModel.sandbox` with `bash`.
+How to actually write the setup and requirement files can be found in the section about [setting up a sandbox](building_blocks/sandbox).
+
+## datasets:
+In the {py:meth}`~columnflow.ml.MLModel.datasets` function, you specify which datasets are important for your machine learning model and which dataset instance(s) should be extracted from your config.
+To use this function your datasets needs to be added to your campaign, as defined by the [Order](https://python-order.readthedocs.io/en/latest/) Module.
+An example can be found [here](https://github.com/uhh-cms/cmsdb/blob/d83fb085d6e43fe9fc362df75bbc12816b68d413/cmsdb/campaigns/run2_2018_nano_uhh_v11/top.py).
+It is recommended to return this as `set`, to prevent double counting.
+In the following example all datasets given by the {ref}`external config <init_of_ml_model>` are taken, but also an additional is given.
+Note how the name of each dataset is used to get a `dataset instance` from your `config instance`.
+This ensures that you properly use the correct dataset.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def datasets
+:end-at: return set(dataset_inst)
+```
+
+
+## produces:
+By default, intermediate columns are not saved within the columnflow framework but are filtered out afterwards.
+If you want to prevent certain columns from being filtered out, you need to tell columnflow.
+You can tell columnsflow, by return the names of all columns that should be preserved, as strings within an interable, in {py:meth}`~columnflow.ml.MLModel.produces`.
+More information can be found in the official documentation about [producers](building_blocks/producers).
+
+In the following example, I want to preserve the number of muons and electrons, we also want to preserve the output of the neural network.
```

Comment:
```suggestion
In the following example, I want tell columnflow to preserve the output of the neural network, but also the fold indices, that are used to create the trainings and test fold.
I do not store the input and target columns of that fold to save disk space, since they are already stored in the training set parquet file.

```